common/core/web/input-processor/src/text/inputProcessor.ts

```diff
@@ -113,14 +117,49 @@ namespace com.keyman.text {
         // If we're performing a 'default command', it's not a standard 'typing' event - don't do fat-finger stuff.
         // Also, don't do fat-finger stuff if predictive text isn't enabled.
         if(this.languageProcessor.isActive && !ruleBehavior.triggersDefaultCommand) {
+          let keyDistribution = keyEvent.keyDistribution;
+
+          // We don't need to track absolute indexing during alternate-generation; 
+          // only position-relative, so it's better to use a sliding window for context
+          // when making alternates.  (Slightly worse for short text, matters greatly
+          // for long text.)
+          let contextWindow = new ContextWindow(preInputMock, ContextWindow.ENGINE_RULE_WINDOW);
+          let windowedMock = contextWindow.toMock();
+
           // Note - we don't yet do fat-fingering with longpress keys.
-          if(keyEvent.keyDistribution && keyEvent.kbdLayer) {
+          if(keyDistribution && keyEvent.kbdLayer) {
+            // Tracks a 'deadline' for fat-finger ops, just in case both context is long enough
+            // and device is slow enough that the calculation takes too long.
+            //
+            // Consider use of https://developer.mozilla.org/en-US/docs/Web/API/Performance/now instead?
+            // Would allow finer-tuned control.
```

Comment:
Yes, let's do this. It's available in Chrome 24+ so should be fine on all platforms we work on.

Otherwise, if we get the tick count with `Date.now()` and we are close to the changeover threshold, we'll get a lot less than 16msec to run.

Reply:
... yay.  `performance.now()` requires use of a `require` statement when in a Node environment (from `perf_hooks`); it's not accessible otherwise.  That's gonna complicate things, as `require` is incompatible with our current toolchain.

Reply:
quick and dirty approach: do we need to time-bound in node environment? response time is less of a factor there.

Reply:
Fair.  I think I figured out a shortcut for that at one point, so that should be possible.

Will continue on this first-thing tomorrow.